### DISCUSSION TITLE
How to set a cookie in a page response? Next appDir

### DISCUSSION THREAD

**Discussion by volnei** [vercel/next.js#43471] [Help]
How to set a cookie in a page response? Next appDir.

Im trying to implement a custom refresh token mechanism and I just can't set a cookie on client side

**Comment by icyJoseph** [comment]
Not sure it is possible, but middleware should work right?

```tsx
import { NextResponse, type NextRequest } from "next/server";

export function middleware(req: NextRequest) {
  const { pathname, search } = req.nextUrl;
  // console.log("Middleware", { pathname, searchj });

  if (pathname === "/pokemon") {
    const response = NextResponse.next();

    response.cookies.set({
      name: "poke-cookie",
      value: `${new Date().getTime()}`,
    });

    return response;
  }

  return NextResponse.next();
}
```

In my case the pokemon route changes search parameters as you type:

```ts
  useEffect(() => {
    if (pathname === null) return;

    router.replace(
      value.trim() ? `${pathname}?pokemon=${value.trim()}` : pathname
    );
  }, [router, pathname, value]);

```

So every keystroke hits the middleware and a new `poke-cookie` is applied:

```shell
Middleware { pathname: '/pokemon', search: '?pokemon=p' }
Middleware { pathname: '/pokemon', search: '?pokemon=pi' }
Middleware { pathname: '/pokemon', search: '?pokemon=pik' }
Middleware { pathname: '/pokemon', search: '?pokemon=pika' }
Middleware { pathname: '/pokemon', search: '?pokemon=pikac' }
Middleware { pathname: '/pokemon', search: '?pokemon=pikach' }
Middleware { pathname: '/pokemon', search: '?pokemon=pikachu' }
```

It sounds like this is the most natural way to do this. You could even handle some kind of redirect logic in the middleware.

```ts

  if (pathname === "/pokemon" && searchParams.get("pokemon") === "redirect") {
    req.nextUrl.pathname = "/";
    req.nextUrl.searchParams.delete("pokemon");
    return NextResponse.redirect(req.nextUrl);
  }
```

**Comment by ephraimduncan** [comment]
Try using the [`cookies-next`](https://github.com/andreizanik/cookies-next) package

Set a cookie on the client side
```js
import { setCookie } from 'cookies-next';

setCookie('key', 'value', options);
```

**Reply by volnei** [reply]
Thank you for your reply! But, client-side doesnt allow http-only cookies :(

**Comment by ayhanap** [comment]
So I am trying to do something similar. My server components do API calls and they use the JWTs that are in cookies. When accessToken expires I refresh tokens with axios interceptor and repeat failed request and then serve the server component.

Now what I cannot do is to return those new tokens to client as cookies. Yeah I can set cookies in middleware but thats only when the request just came in, I want to swt cookies before the response about to be sent back, that when my server component rendered all with that refresh token etc.
